crates/turbo-tasks-memory/src/task.rs

```diff
@@ -92,15 +106,160 @@ pub struct Task {
     /// The type of the task
     ty: TaskType,
     /// The mutable state of the task
-    state: RwLock<TaskState>,
+    /// Unset state is equal to a Dirty task that has not been executed yet
+    state: RwLock<TaskMetaState>,
+}
+
+enum TaskMetaState {
+    Full(Box<TaskState>),
+    Partial(Box<PartialTaskState>),
+    Unloaded(UnloadedTaskState),
+}
+
+impl Default for TaskMetaState {
+    fn default() -> Self {
+        Self::Unloaded(UnloadedTaskState::default())
+    }
+}
+
+impl TaskMetaState {
+    fn into_unwrap_partial(self) -> PartialTaskState {
+        match self {
+            Self::Partial(state) => *state,
+            _ => panic!("TaskMetaState is not partial"),
+        }
+    }
+
+    fn into_unwrap_unloaded(self) -> UnloadedTaskState {
+        match self {
+            Self::Unloaded(state) => state,
+            _ => panic!("TaskMetaState is not none"),
+        }
+    }
+
+    fn unwrap_full(&self) -> &TaskState {
```

Comment:
Nit: These I would call `as_full_unwrap` or `unwrap_as_full` (for `tms.as_full().unwrap()`)